packages/langium/src/grammar/validation/types-validator.ts

```diff
@@ -67,6 +77,31 @@ export class LangiumGrammarTypesValidator {
 
 ///////////////////////////////////////////////////////////////////////////////
 
+function isCyclicType(type: ast.TypeDefinition | ast.AbstractType, visited: Set<AstNode>): boolean {
+    if (visited.has(type)) {
+        return true;
+    }
+    visited.add(type);
+    if (ast.isSimpleType(type)) {
+        if (type.typeRef?.ref) {
+            return isCyclicType(type.typeRef!.ref, visited);
+        }
+    } else if (ast.isReferenceType(type)) {
+        return isCyclicType(type.referenceType, visited);
+    } else if (ast.isArrayType(type)) {
+        return isCyclicType(type.elementType, visited);
+    } else if (ast.isUnionType(type)) {
+        for (const t of type.types) {
+            if (isCyclicType(t, visited)) {
+                return true;
+            }
+        }
```

Comment:
Suggestion: Since we return with `false` anyway after the loop, we can immediately return here.

```suggestion
        return type.types.some(t => isCyclicType(t, visited));
```